src/inc/til/rect.h

```diff
@@ -0,0 +1,856 @@
+        constexpr SMALL_RECT to_small_rect() const
+        {
+            // The two -1 operations below are technically UB if they underflow.
+            // But practically speaking no hardware without two's complement for
+            // signed integers is supported by Windows. If they do underflow, they'll
+            // result in INT_MAX which will throw in gsl::narrow just like INT_MAX does.
+            return {
+                gsl::narrow<short>(left),
+                gsl::narrow<short>(top),
+                gsl::narrow<short>(right - 1),
+                gsl::narrow<short>(bottom - 1),
+            };
+        }
+#endif
+
+#ifdef _WINDEF_
+        explicit constexpr rect(const RECT& other) noexcept :
+            rect{ other.left, other.top, other.right, other.bottom }
+        {
+        }
+
+        constexpr RECT to_win32_rect() const noexcept
+        {
+            return { left, top, right, bottom };
+        }
+#endif
+
+#ifdef DCOMMON_H_INCLUDED
+        template<typename TilMath>
+        constexpr rect(TilMath&& math, const D2D1_RECT_F& other) :
+            rect{ std::forward<TilMath>(math), other.left, other.top, other.right, other.bottom }
+        {
+        }
+
+        constexpr D2D1_RECT_F to_d2d_rect() const noexcept
+        {
+            return {
+                static_cast<float>(left),
+                static_cast<float>(top),
+                static_cast<float>(right),
+                static_cast<float>(bottom),
+            };
+        }
+#endif
+
+#ifdef WINRT_Windows_Foundation_H
+        template<typename TilMath>
+        constexpr rect(TilMath&& math, const winrt::Windows::Foundation::Rect& other) :
+            rect{ std::forward<TilMath>(math), other.X, other.Y, other.X + other.Width, other.Y + other.Height }
+        {
+        }
+
+        winrt::Windows::Foundation::Rect to_winrt_rect() const noexcept
+        {
+            return {
+                static_cast<float>(left),
+                static_cast<float>(top),
+                static_cast<float>(width()),
+                static_cast<float>(height()),
+            };
+        }
+#endif
+
+#ifdef WINRT_Microsoft_Terminal_Core_H
+        template<typename TilMath>
+        constexpr rect(TilMath&& math, const winrt::Microsoft::Terminal::Core::Padding& other) :
+            rect{ std::forward<TilMath>(math), other.Left, other.Top, other.Right, other.Bottom }
+        {
+        }
+
+        winrt::Microsoft::Terminal::Core::Padding to_core_padding() const noexcept
+        {
+            return {
+                static_cast<float>(left),
+                static_cast<float>(top),
+                static_cast<float>(right),
+                static_cast<float>(bottom),
+            };
+        }
+#endif
+
+        std::wstring to_string() const
+        {
+            return wil::str_printf<std::wstring>(L"(L:%d, T:%d, R:%d, B:%d) [W:%d, H:%d]", left, top, right, bottom, width(), height());
+        }

```

Comment:
Is this used outside of testing scenarios?

Reply:
I renamed the file and this is the original code (almost everything in rect.h is unmodified!).
I'll add some ifdefs around this. 🙂

Reply:
Oh sure. Feel free to ignore me then.